backend/invoke_ai_web_server.py

```diff
@@ -611,23 +621,41 @@ def image_done(image, seed, first_seed):
                 if self.canceled.is_set():
                     raise CanceledException
 
-                if gfpgan_parameters:
-                    progress.set_current_status('Restoring Faces')
+                if facetool_parameters:
+                    if facetool_parameters['type'] == 'gfpgan':
+                        progress.set_current_status('Restoring Faces (GFPGAN)')
+                    elif facetool_parameters['type'] == 'codeformer':
+                        progress.set_current_status('Restoring Faces (Codeformer)')
+
                     progress.set_current_status_has_steps(False)
                     self.socketio.emit(
                         'progressUpdate', progress.to_formatted_dict()
                     )
                     eventlet.sleep(0)
 
-                    image = self.gfpgan.process(
-                        image=image,
-                        strength=gfpgan_parameters['strength'],
-                        seed=seed,
-                    )
+                    if facetool_parameters['type'] == 'gfpgan':
+                        image = self.gfpgan.process(
+                            image=image,
+                            strength=facetool_parameters['strength'],
+                            seed=seed,
+                        )
+                    elif facetool_parameters['type'] == 'codeformer':
+                        image = self.codeformer.process(
+                            image=image,
+                            strength=facetool_parameters['strength'],
+                            fidelity=facetool_parameters['codeformer_fidelity'],
+                            seed=seed,
+                            device='cpu' if str(self.generate.device) == 'mps' else self.generate.device,
```

Comment:
FMI - Why is this here for codeformer but not for GFPGAN?

Reply:
I suppose it's needed bc codeformer can run on CUDA but not MPS, while GFPGAN can run on any device (or always on CPU - unsure). Doesn't work if you don't provide the device though.